Add `NoAudienceBannerWidget` component (+additional refactoring)

kuasha420 opened the pull request:
## Summary

Addresses issue:

- #8155

## Relevant technical choices

1. The implementation mostly follows the IB with a few exceptions that are detailed below.
2. This does not implement the banner for view only context as those will be handled in separate issues. 
3. Introduces a new `LeanCTABanner` that is used for the `NoAudienceBannerWidget`. 
4. Refactors `ConnectAnalyticsCTAWidget` to use the aforementioned `LeanCTABanner` since their design is mostly identical.
5. SVG graphics sizes are standardized (for the most part) across both banners. We should inform @sigal-teller of this so future banners of similar kind follow a consistent size and positioning, keeping things consistent and preventing reinvention of wheels.  
6. A few components changed location to fit the new file structure. 
7. Feedback welcome!

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

benbowler commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget/index.test.js:
```diff
@@ -0,0 +1,259 @@
+		'analyticsAudienceTiles'
+	);
+
+	const auduenceSettingsRegExp = new RegExp(
```
I was in here while writing an IB for another ticket and noticed a typo here. auduence -> audience

nfmohit replied:
Thanks @benbowler ! I had noticed this and have included this in the CR feedback.

nfmohit commented on assets/sass/components/dashboard/_googlesitekit-lean-cta-banner.scss:
The changes here look good. However, I've proposed some changes to the class names to adhere to the BEM principles in my former comments. Thanks!

nfmohit commented on assets/sass/components/dashboard/_googlesitekit-lean-cta-banner.scss:
```diff
@@ -67,21 +67,13 @@
 				}
 
-				.googlesitekit-audience-connect-analytics-cta__text {
+				.googlesitekit-lean-cta-banner--body__text {
```
I've also proposed removing this class name and applying its styles to all child paragraph tags of the body, as it seems this class name is being applied to all paragraph tags anyway. Let me know what you think, thank you!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/NoAudienceBannerWidget/index.js:
```diff
@@ -0,0 +1,77 @@
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import whenActive from '../../../../../../util/when-active';
```
Let's move this group to the "_Internal dependencies_" group.

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/NoAudienceBannerWidget/index.js:
```diff
@@ -0,0 +1,77 @@
+import NoAudienceBanner from './NoAudienceBanner';
+
+function NoAudienceBannerWidget( { Widget, WidgetNull } ) {
+	const availableAudiences = useSelect( ( select ) => {
+		const audiences = select( MODULES_ANALYTICS_4 ).getAvailableAudiences();
+		return audiences?.map( ( audience ) => audience.name );
+	} );
+	const configuredAudiences = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getConfiguredAudiences()
+	);
```
I think both of these should use the `useInViewSelect` hook as they are in the widget area and their resolvers will make API requests. What do you think?

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/NoAudienceBannerWidget/NoAudienceBanner.js:
```diff
@@ -0,0 +1,130 @@
+		<LeanCTABanner
+			className="googlesitekit-no-audience-banner"
+			Icon={ Icon }
+			SVGGraphic={ NoAudienceBannerGraphic }
```
It seems the graphic here renders quite larger than what we have in Figma, thus making the entire widget quite larger as well. Should we limit its height in the respective stylesheet?

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/NoAudienceBannerWidget/NoAudienceBanner.js:
```diff
@@ -0,0 +1,130 @@
+			Icon={ Icon }
+			SVGGraphic={ NoAudienceBannerGraphic }
+		>
+			<p className="googlesitekit-lean-cta-banner--body__text">
```
In my former comments, I've added a proposal to remove this class name.

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/NoAudienceBannerWidget/NoAudienceBanner.js:
```diff
@@ -0,0 +1,130 @@
+						}
+					) }
+			</p>
+			<p className="googlesitekit-lean-cta-banner--body__text">
```
Same here. I've added a proposal to remove this class name in my former comments.

nfmohit commented on assets/sass/components/analytics-4/audience-segmentation/_googlesitekit-no-audience-banner.scss:
```diff
@@ -0,0 +1,32 @@
+
+	.googlesitekit-lean-cta-banner.googlesitekit-no-audience-banner {
+
+		.googlesitekit-lean-cta-banner--graphic {
```
According to my former proposal to change this class name to meet BEM principles:

```suggestion
		.googlesitekit-lean-cta-banner__graphic {
```

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/NoAudienceBannerWidget/index.test.js:
```diff
@@ -0,0 +1,297 @@
+		'analyticsNoAudienceBanner'
+	);
+
+	const auduenceSettingsRegExp = new RegExp(
```
```suggestion
	const audienceSettingsRegExp = new RegExp(
```

nfmohit reviewed (changes requested):
Brilliant work here, thank you @kuasha420! I have left some comments for your purview. Please let me know what you think, thank you!

kuasha420 commented:
@nfmohit Thank you for the excellent CR feedback and giving me a refresher on the BEM principles. I've addressed your feedback. Additionally, I've made the `InfoNoticeWidget` only render when there is matching audiences, as discussed [here](https://10up.slack.com/archives/C079RDV88AZ/p1719404713021239). I have added a point in QAB and ACs to mention that. Cheers.

nfmohit reviewed (changes requested):
Thanks so much for addressing my feedback and updating the PR, @kuasha420 ! The code looks excellent to me. However, while testing this further, I've noticed a few quirks:

#### Observation 1

This is very intermittent but I've noticed that the Audience Segmentation widget area doesn't show up sometimes. See the following:

![1](https://github.com/user-attachments/assets/8dfab5be-54e4-4f13-8e5f-2c184c1248dc)

This is what the markup looks like:
![image](https://github.com/user-attachments/assets/24fab7ef-0cf4-43c3-8095-05ab289678e4)

I'm not a 100% sure why this happens, but based on my short-lived investigation, it appears to not occur when we use `useSelect` instead of `useInViewSelect` to look up `availableAudiences` and `configuredAudiences` to determine the visibility conditions for these widgets. Could this be because the widgets return `null` if those selectors do not resolve, and they do not resolve because they don't come into view? Could you investigate if that is the case? If so, I think we should use `useSelect` in this context as these determine if the widgets should actually appear.

#### Observation 2

The banner graphic appears to have some padding on its horizontal axis. See screenshot from Figma:

![image](https://github.com/user-attachments/assets/04d12e9c-f24e-4d2d-bd62-240ccbf214e0)

This is specially visible in the "no audiences available" state where the text pushes the graphic fully to the right. See:

![image](https://github.com/user-attachments/assets/e1fc05e0-0a63-4e2e-a849-69cbe3df4042)

-------------------------

Please let me know what you think of the above, and if you have any questions. Thank you!

nfmohit reviewed (approved):
LGTM 👍 Thank you for the excellent work and all the patience on this, @kuasha420 !
